lib/Runtime/Types/DeferredTypeHandler.cpp

```diff
@@ -24,7 +24,26 @@ namespace Js
 
         ScriptContext* scriptContext = instance->GetScriptContext();
         instance->EnsureSlots(0, typeHandler->GetSlotCapacity(), scriptContext, typeHandler);
-        typeHandler->SetInstanceTypeHandler(instance);
+
+        FunctionProxy * functionProxy = instance->GetFunctionProxy();
+        ScriptFunctionType * undeferredFunctionType = nullptr;
+        if (functionProxy)
+        {
+            undeferredFunctionType = functionProxy->GetUndeferredFunctionType();
+        }
+        if (undeferredFunctionType && undeferredFunctionType->GetTypeHandler()->GetMayBecomeShared())
+        {
+            instance->ReplaceType(undeferredFunctionType);
+            instance->ShareType();
+        }
+        else
+        {
+            typeHandler->SetInstanceTypeHandler(instance);
+            if (functionProxy && typeHandler->GetMayBecomeShared())
+            {
+                functionProxy->SetUndeferredFunctionType(ScriptFunction::FromVar(instance)->GetScriptFunctionType());
```

Comment:
>ScriptFunction::FromVar(instance) [](start = 57, length = 33)

How do we know that instance is a ScriptFunction? (Not saying it's wrong, I'm just curious.)

Reply:
ScriptFunction's are the only JavascriptFunction's with non-null FunctionProxy.

Reply:
Could use ScriptFunction::UnsafeFromVar.

---
In reply to: [171357540](https://github.com/Microsoft/ChakraCore/pull/4748#discussion_r171357540) [](ancestors = 171357540)

Reply:
Done. Thanks.